File: java/src/main/java/com/hindu/calendar/core/Tithi.java
package com.hindu.calendar.core;

import com.hindu.calendar.ephemeris.Ephemeris;
import com.hindu.calendar.model.Location;
import com.hindu.calendar.model.Paksha;
import com.hindu.calendar.model.TithiInfo;

public class Tithi {

    private final Ephemeris ephemeris;

    public Tithi(Ephemeris ephemeris) {
        this.ephemeris = ephemeris;
    }

    public double lunarPhase(double jdUt) {
        double moon = ephemeris.lunarLongitude(jdUt);
        double sun = ephemeris.solarLongitude(jdUt);
        double phase = (moon - sun) % 360.0;
        if (phase < 0) phase += 360.0;
        return phase;
    }

    public int tithiAtMoment(double jdUt) {
        double phase = lunarPhase(jdUt);
        int t = (int) (phase / 12.0) + 1;
        if (t > 30) t = 30;
        return t;
    }

    public double findTithiBoundary(double jdStart, double jdEnd, int targetTithi) {
        double targetPhase = (targetTithi - 1) * 12.0;

        double lo = jdStart;
        double hi = jdEnd;

        for (int i = 0; i < 50; i++) {
            double mid = (lo + hi) / 2.0;
            double phase = lunarPhase(mid);

            double diff = phase - targetPhase;
            if (diff > 180.0) diff -= 360.0;
            if (diff < -180.0) diff += 360.0;

            if (diff >= 0)
                hi = mid;
            else
                lo = mid;
        }

        return (lo + hi) / 2.0;
    }

    public TithiInfo tithiAtSunrise(int year, int month, int day, Location loc) {
        double jd = ephemeris.gregorianToJd(year, month, day);
        double jdRise = ephemeris.sunriseJd(jd, loc);

        if (jdRise <= 0) {
            jdRise = jd + 0.5 - loc.utcOffset() / 24.0;
        }

        double riseUt = jdRise;
        int t = tithiAtMoment(riseUt);

        Paksha paksha = (t <= 15) ? Paksha.SHUKLA : Paksha.KRISHNA;
        int pakshaTithi = (t <= 15) ? t : t - 15;

        double jdStart = findTithiBoundary(riseUt - 2.0, riseUt, t);

        int nextTithi = (t % 30) + 1;
        double jdEnd = findTithiBoundary(riseUt, riseUt + 2.0, nextTithi);

        // Check for kshaya tithi
        boolean isKshaya = false;
        double jdTomorrow = jd + 1.0;
        double jdRiseTmrw = ephemeris.sunriseJd(jdTomorrow, loc);
        if (jdRiseTmrw > 0) {
            int tTmrw = tithiAtMoment(jdRiseTmrw);
            int diff = ((tTmrw - t) % 30 + 30) % 30;
            isKshaya = (diff > 1);
        }

        return new TithiInfo(t, paksha, pakshaTithi, jdStart, jdEnd, isKshaya);
    }
}

File: rust/src/core/tithi.rs
use crate::ephemeris::Ephemeris;
use crate::model::*;

pub fn lunar_phase(eph: &mut Ephemeris, jd_ut: f64) -> f64 {
    let moon = eph.lunar_longitude(jd_ut);
    let sun = eph.solar_longitude(jd_ut);
    let mut phase = (moon - sun) % 360.0;
    if phase < 0.0 { phase += 360.0; }
    phase
}

pub fn tithi_at_moment(eph: &mut Ephemeris, jd_ut: f64) -> i32 {
    let phase = lunar_phase(eph, jd_ut);
    let t = (phase / 12.0) as i32 + 1;
    if t > 30 { 30 } else { t }
}

pub fn find_tithi_boundary(
    eph: &mut Ephemeris,
    jd_start: f64,
    jd_end: f64,
    target_tithi: i32,
) -> f64 {
    let target_phase = (target_tithi - 1) as f64 * 12.0;
    let mut lo = jd_start;
    let mut hi = jd_end;

    for _ in 0..50 {
        let mid = (lo + hi) / 2.0;
        let phase = lunar_phase(eph, mid);
        let mut diff = phase - target_phase;
        if diff > 180.0 { diff -= 360.0; }
        if diff < -180.0 { diff += 360.0; }

        if diff >= 0.0 {
            hi = mid;
        } else {
            lo = mid;
        }
    }

    (lo + hi) / 2.0
}

pub fn tithi_at_sunrise(
    eph: &mut Ephemeris,
    year: i32,
    month: i32,
    day: i32,
    loc: &Location,
) -> TithiInfo {
    let jd = eph.gregorian_to_jd(year, month, day);
    let mut jd_rise = eph.sunrise_jd(jd, loc);

    if jd_rise <= 0.0 {
        jd_rise = jd + 0.5 - loc.utc_offset / 24.0;
    }

    let t = tithi_at_moment(eph, jd_rise);

    let paksha = if t <= 15 { Paksha::Shukla } else { Paksha::Krishna };
    let paksha_tithi = if t <= 15 { t } else { t - 15 };

    let jd_start = find_tithi_boundary(eph, jd_rise - 2.0, jd_rise, t);

    let next_tithi = (t % 30) + 1;
    let jd_end = find_tithi_boundary(eph, jd_rise, jd_rise + 2.0, next_tithi);

    // Check for kshaya tithi
    let jd_tomorrow = jd + 1.0;
    let jd_rise_tmrw = eph.sunrise_jd(jd_tomorrow, loc);
    let is_kshaya = if jd_rise_tmrw > 0.0 {
        let t_tmrw = tithi_at_moment(eph, jd_rise_tmrw);
        let diff = ((t_tmrw - t) + 30) % 30;
        diff > 1
    } else {
        false
    };

    TithiInfo {
        tithi_num: t,
        paksha,
        paksha_tithi,
        jd_start,
        jd_end,
        is_kshaya,
    }
}
